snuba/utils/streams/abstract.py

```diff
@@ -0,0 +1,137 @@
+# representation (after deserialization of the raw bytes) is generally more
+# appropriate.
+TValue = TypeVar("TValue")
+
+
+@dataclass
+class Message(Generic[TStream, TOffset, TValue]):
+    """
+    Represents a single message within a stream.
+    """
+
+    stream: TStream
+    offset: TOffset
+    value: TValue
+
+
+class ConsumerError(Exception):
+    """
+    Base class for exceptions that are raised during consumption.
+
+    Subclasses may extend this class to disambiguate errors that are specific
+    to their implementation.
+    """
+
+
+class EndOfStream(ConsumerError, Generic[TStream]):
+    """
+    Raised when there are no more messages to consume from the stream.
+    """
+
+    def __init__(self, stream: TStream):
+        self.stream = stream
+
+
+class Consumer(ABC, Generic[TStream, TOffset, TValue]):
+    """
+    An abstract class that provides methods to consume messages from a
+    multiplexed collection of streams.
+
+    This interface is heavily "inspired" by the Confluent Kafka Consumer
+    implementation, but only exposes a limited set of the available methods
+    and several method signatures and/or return values differ -- these
+    changes are intended to improve the ease of use of the consumer, as well
+    as make development and testing more straightforward. Most code that uses
+    the Confluent driver can be modified to use this interface with minimal
+    changes, but it is not intended to be a drop-in replacement for the
+    Confluent implementation.
+
+    All methods are blocking unless otherise noted in the documentation for
+    that method.
+    """
+
+    @abstractmethod
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[TStream]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[TStream]], None]] = None,
+    ) -> None:
+        """
+        Subscribe to topic streams. This replaces a previous subscription.
+        This method does not block. The subscription may not be fulfilled
+        immediately: instead, the ``on_assign`` and ``on_revoke`` callbacks
+        are called when the subscription state changes with the updated
+        assignment for this consumer.
+
+        Raises a ``RuntimeError`` if called on a closed consumer.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def poll(
+        self, timeout: Optional[float] = None
+    ) -> Optional[Message[TStream, TOffset, TValue]]:
+        """
+        Return the next message available to be consumed, if one is
+        available. If no message is available, this method will block up to
+        the ``timeout`` value before returning ``None``. A timeout of
+        ``0.0`` represents "do not block", while a timeout of ``None``
+        represents "block until a message is available (or forever)".
+
+        Calling this method may also invoke subscription state change
+        callbacks.
```

Comment:
Do you mean exclusively for the Kafka implementation (and removing it here) or to duplicate this comment there as well?

Reply:
I meant exclusively for the Kafka implementation, since, is I understand it correctly, that's a quirk of how the Kafka implementation works.

Reply:
I think we'd like to emulate this same behavior in other consumers — they should always invoke the subscription callback at least once, even if they aren't using consumer groups or any sort of coordination. For example the file one we talked about yesterday should probably fire the callback on the first `poll` call with the list of files that match the `topics` (note: might make sense to use a more abstract word than "topic"? not sure) provided to the `subscribe` call. Related, I'd like to abstract the `KafkaConsumer` test soon so that we can ensure all implementations of `Consumer` behave the same, I just haven't got that far yet.